x/coin.go

```diff
@@ -75,22 +75,54 @@ func (c Coin) Divide(pieces int64) (Coin, Coin, error) {
 	return one, rest, nil
 }
 
-// Multiply returns the result of a coin value multiplication.
-func (c Coin) Multiply(times int64) Coin {
-	whole := c.Whole * times
-	frac := c.Fractional * times
+// Multiply returns the result of a coin value multiplication. This method can
+// fail if the result would overflow maximum coin value.
+func (c Coin) Multiply(times int64) (Coin, error) {
+	if times == 0 || (c.Whole == 0 && c.Fractional == 0) {
+		return Coin{Ticker: c.Ticker}, nil
+	}
+
+	whole, ok := mul64(c.Whole, times)
+	if !ok {
+		return Coin{}, errors.ErrOverflow
+
+	}
+	frac, ok := mul64(c.Fractional, times)
+	if !ok {
+		return Coin{}, errors.ErrOverflow
+	}
 
 	// Normalize if fractional value overflows.
 	if frac > FracUnit {
-		whole += frac / FracUnit
+		if n := whole + frac/FracUnit; n < whole {
```

Comment:
Nice check here... gotta have a clever test that multiply doesn't overflow until we carry the fractional, but good to cover this edge

Reply:
This is tested by https://github.com/iov-one/weave/pull/351/files#diff-2a10d0bf5e4f9e87e59d8885f76a0278R399